blogposts/2021/postgres-version-update.md

```diff
@@ -0,0 +1,139 @@
+
+So, instead, we should count fewer things.
+
+In our naïve attempt to determine the progress of a particular migration, we wanted to simply report the ratio between the number of rows with a version _above_ our target and the total number of rows in the table. However, we don't need progress reporting to be completely accurate (except for at the 0% and 100% edges).
+
+Each table we care to migrate in the codeintel database has a composite primary key containing an _index identifier_. Each unique index identifier corresponds to an LSIF index file uploaded by a user.
+
+Instead of counting the ratio between migrated _rows_ and total rows (which is very large), we can instead count the ratio between migrated _indexes_ and total indexes (which is orders of magnitude smaller - 300 million vs. 23 thousand).
+
+The **obvious approach** would be to migrate all rows in a table associated with the same index at once. Unfortunately, indexes can be very massive. Some enterprise customers are regularly uploading 13GB index files. Thus, our migration batch size is likely going to be too small to guarantee all rows can be migrated, and increasing the batch size will add additional memory and compute constraints on the migrating container.
+
+The **approach we chose** tracks the minimum and maximum row versions for each index in a separate, _much_ smaller table.
+
+*lsif\_data\_definitions*:
+
+| index_id | scheme | identifier | data | version | 
+| -------- | ------ | ---------- | ---- | ------- | 
+| 1        | foo    | bar        | .... | 1       | 
+| 1        | bar    | baz        | .... | 2       | 
+| 2        | baz    | quux       | .... | 2       | 
+| 2        | quux   | bonk       | .... | 2       |
+| 3        | bonk   | honk       | .... | 1       |
+
+*lsif\_data\_definitions\_schema\_versions*:
+
+| index_id | min_version | max_version |
+| -------- | ----------- | ----------- |
+| 1        | 1           | 2           |
+| 2        | 2           | 2           |
+| 3        | 1           | 1           |
+
+This second table allows us to efficiently determine which indexes have been migrated and which still need attention. For example, if we are currently migrating from schema version 1 to 2, then the table above shows index 1 is partially migrated, index 2 is fully migrated, and index 3 has yet to be touched.
+
+It is an efficient operation to count both the total number of indexes as well as the number of indexes that are above or below a particular schema version.
+
+The schema version bounds can be kept in sync with the source table trivially by using triggers on inserts, updates, and deletes: if we insert a new row with a new lowest or new highest version or if we update/delete a row with the last remaining lowest or highest version, then we update the associated minimum or maximum version for that index.
+
+### The catch
```

Comment:
This section could use a single summary sentence of what "the catch" is. If this already exists and I'm just missing it, you can disregard this comment. Alternatively, you can rename this section to be more descriptive.